### DISCUSSION TITLE
New Feature Alert: Integrate use of Memory in Planner Skill

### DISCUSSION THREAD

**Discussion by lemillermicrosoft** [microsoft/semantic-kernel#170] [General]
Hello everyone,

We're thrilled to announce a new feature in the Semantic Kernel's planner skill that will empower developers to create even more powerful applications. With the latest update, developers can now use embeddings exposed in the kernel via `SemanticTextMemory` to determine related skills, significantly enhancing the functionality and usability of the planner skill.  

### Improvements
These features help the AI "Goal Seek" better for you.  

#### Token Limits
In this example using GPT-3 our previous Planner fails to produce a plan when using all of the sample skills because the token limit is hit very quickly.

```
<goal>
Write a short story about a developer testing some code, translate to french, and then e-mail it to Amir.
</goal>
Error: InvalidRequest: The request is not valid, HTTP status: BadRequest
```

Using semantic memories we're able to give the Planner improved knowledge helping it provide a complete and improved plan such as:
```
<goal>
Write a short story about a developer testing some code, translate to french, and then e-mail it to Amir.
</goal>
<plan>
  <function.WriterSkill.ShortStory input="A developer testing some code" />
  <function.WriterSkill.Translate language="French" setContextVariable="TRANSLATED_STORY" />
  <function.email.GetEmailAddress input="Amir" setContextVariable="EMAIL_ADDRESS" />
  <function.email.SendEmail input="$TRANSLATED_STORY" email_address="$EMAIL_ADDRESS" />
</plan>
```

#### Semantic Results
With GPT4, token limit is less of an issue, but the impact of a large manual can still be seen in this example using the previous Planner. Note that the LLM is unable to connect the output of the Joke function to the DOSScript function properly.

```
<goal>
Write a joke about how programming is hard and then write a dos script to output the joke to the screen. My name is Bart.
</goal>
<plan>
  <function.FunSkill.Joke style="programming" setContextVariable="JOKE" />
  <function.CodingSkill.DOSScript input="output the joke to the screen" firstName="Bart" />
</plan>
```

Using the new Planner capabilities to lower the size of the manual based on the goal, you can see the plan produced is very similar, but now correctly passes input to DOSScript function:
```
<goal>
Write a joke about how programming is hard and then write a dos script to output the joke to the screen. My name is Bart.
</goal>
<plan>
  <function.FunSkill.Joke style="programming" setContextVariable="JOKE"/>
  <function.CodingSkill.DOSScript firstName="Bart" input="$JOKE" appendToResult="RESULT__DOS_SCRIPT"/>   
</plan>
```

### How
To take advantage of this powerful new feature, developers simply need to update their Semantic Kernel package to the latest version [0.8.56.1-preview](https://www.nuget.org/packages/Microsoft.SemanticKernel/0.8.56.1-preview), configure their kernel instance with an embeddings backend and an IMemoryStore implementation (such as the built-in VolatileMemoryStore), and configure the new parameters and options as needed when calling the CreatePlan PlannerSkill function.
This level of customization allows developers to fine-tune their applications to meet their specific needs and objectives, while also providing users with the most relevant functions to execute via natural language queries.

Here are some code snippets to illustrate how to take advantage of the new features:

Build your kernel with Embeddings backend and an associated MemoryStore.
```
var kernel = new KernelBuilder()
    .Configure(
        config =>
        {
            config.AddAzureOpenAICompletionBackend(...);
            config.AddAzureOpenAIEmbeddingsBackend(...);
        })
    .WithMemoryStorage(new VolatileMemoryStore())
    .Build();
```

Import skills as before
```
var planner = kernel.ImportSkill(new PlannerSkill(kernel), "planning");
kernel.ImportSkill(new EmailSkill(), "email");
kernel.ImportSkill(new TextSkill(), "text");
// import any other skills or functions ....
```

Create your plan using relevancy or other options:
```
var context = new ContextVariables("Create a book with 3 chapters about a group of kids in a club called 'The Thinking Caps.'");

// Set the RelevancyThreshold to enable semantic filtering
context.Set(PlannerSkill.Parameters.RelevancyThreshold, "0.78");
// context.Set(PlannerSkill.Parameters.MaxRelevantFunctions, "5");

// To exclude certain functions or entire Skills from being included in plan creation requests
// context.Set(PlannerSkill.Parameters.ExcludedFunctions, "NovelChapter");
// context.Set(PlannerSkill.Parameters.ExcludedSkills, "email");

// Or to ensure specific functions are included in plan creation requests regardless of semantic filtering
// context.Set(PlannerSkill.Parameters.IncludedFunctions, "DadJoke");

var results = await kernel.RunAsync(context, planner["CreatePlan"]);
```

### Feedback
We think these capabilities will improve your results as you leverage the planner to help you accomplish tasks in your applications and would love to hear your feedback.

Thanks,
Lee

**Comment by eolasd** [comment]
This is great. Is is possible to have the planner use chat gpt 3.5 ( since it has much lower cost) instead of davinci-003?
